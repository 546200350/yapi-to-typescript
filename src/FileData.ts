export { FileData as default } from 'vtils'
