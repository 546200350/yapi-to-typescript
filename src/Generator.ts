import JSON5 from 'json5'
import { CategoryList, Config, Interface, InterfaceList, Method, PropDefinition, RequestBodyType, RequestFormItemType, Required, ResponseBodyType } from './types'
import { isEmpty } from 'vtils'
import { JSONSchema4 } from 'json-schema'
import { jsonSchemaStringToJsonSchema, jsonSchemaToType, jsonToJsonSchema, mockjsTemplateToJsonSchema, propDefinitionsToJsonSchema } from './utils'

export default class Generator {
  /** 配置 */
  private config: Config

  /** { 项目标识: 分类列表 } */
  private projectIdToCategoryList: Record<string, CategoryList | undefined> = Object.create(null)

  constructor(config: Config) {
    this.config = config
  }

  /** 生成请求数据类型 */
  static async generateRequestDataType(
    { interfaceInfo, typeName }: {
      interfaceInfo: Interface,
      typeName: string,
    },
  ): Promise<string> {
    let jsonSchema: JSONSchema4 = {}

    switch (interfaceInfo.method) {
      case Method.GET:
      case Method.HEAD:
      case Method.OPTIONS:
        jsonSchema = propDefinitionsToJsonSchema(
          interfaceInfo.req_query.map<PropDefinition>(item => ({
            name: item.name,
            required: item.required === Required.true,
            type: 'string',
            comment: item.desc,
          })),
        )
        break
      default:
        switch (interfaceInfo.req_body_type) {
          case RequestBodyType.form:
            jsonSchema = propDefinitionsToJsonSchema(
              interfaceInfo.req_body_form.map<PropDefinition>(item => ({
                name: item.name,
                required: item.required === Required.true,
                type: (item.type === RequestFormItemType.file ? 'file' : 'string') as any,
                comment: item.desc,
              })),
            )
            break
          case RequestBodyType.json:
            if (interfaceInfo.req_body_other) {
              jsonSchema = interfaceInfo.req_body_is_json_schema
                ? jsonSchemaStringToJsonSchema(interfaceInfo.req_body_other)
                : jsonToJsonSchema(JSON5.parse(interfaceInfo.req_body_other))
            }
            break
          default:
            break
        }
        break
    }

    return jsonSchemaToType(jsonSchema, typeName)
  }

  /** 生成响应数据类型 */
  static async generateResponseDataType(
    { interfaceInfo, typeName, dataKey }: {
      interfaceInfo: Interface,
      typeName: string,
      dataKey?: string,
    },
  ): Promise<string> {
    let jsonSchema: JSONSchema4 = {}

    switch (interfaceInfo.res_body_type) {
      case ResponseBodyType.json:
      case ResponseBodyType.jsonSchema:
        if (interfaceInfo.res_body) {
          jsonSchema = interfaceInfo.res_body_is_json_schema
            ? jsonSchemaStringToJsonSchema(interfaceInfo.res_body)
            : mockjsTemplateToJsonSchema(JSON5.parse(interfaceInfo.res_body))
        }
        break
      default:
        break
    }

    if (dataKey && jsonSchema && jsonSchema.properties && jsonSchema.properties[dataKey]) {
      jsonSchema = jsonSchema.properties[dataKey]
    }

    return jsonSchemaToType(jsonSchema, typeName)
  }

  /** 获取分类的接口列表 */
  async fetchCategoryInterfaceList(
    { serverUrl, token, categoryId }: {
      serverUrl: string,
      token: string,
      categoryId: number,
    },
  ): Promise<InterfaceList> {
    const projectId: string = `${serverUrl}|${token}`

    if (!(projectId in this.projectIdToCategoryList)) {
      this.projectIdToCategoryList[projectId] = []
    }

    const category = (this.projectIdToCategoryList[projectId] || []).find(
      cat => (
        !isEmpty(cat)
          && !isEmpty(cat.list)
          && cat.list[0].catid === categoryId
      ),
    )

    return category ? category.list : []
  }
}
