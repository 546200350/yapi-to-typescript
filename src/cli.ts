#!/usr/bin/env node

import { run } from './cliCore'

(async () => {
  await run()
})()
