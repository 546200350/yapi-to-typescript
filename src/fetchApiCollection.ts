import request from 'superagent'
import { ApiCollection, Config } from './types'

type LoginResult = {
  body: {
    errcode: number,
    errmsg: string,
  },
}
type ApiResult = { body: Buffer }

export default async function fetchApiCollection(config: Config): Promise<ApiCollection> {
  const [, baseUrl, projectId] = config.projectUrl.match(/(.+\/)project\/(\d+)\//)

  const loginUrl = `${baseUrl}api/user/${config.login.method === 'ldap' ? 'login_by_ldap' : 'login'}`
  const apiUrl = `${baseUrl}api/plugin/export?type=json&pid=${projectId}&status=all&isWiki=false`

  const agent = request.agent()

  if (config.extraCookies) {
    (agent as any).set('Cookie', config.extraCookies)
  }

  const { body: loginResult } = await agent
    .post(loginUrl)
    .send({
      email: config.login.email,
      password: config.login.password,
    }) as LoginResult
  if (!loginResult || loginResult.errcode !== 0) {
    throw new Error(
      `\x1b[31m登录 ${loginUrl} 失败，请检查邮箱、密码是否有误或服务是否可用。（服务器错误信息：${loginResult.errmsg}）\x1b[0m`
    )
  }

  const { body: api } = await agent.get(apiUrl).buffer() as ApiResult

  return JSON.parse(api.toString()) as any
}
