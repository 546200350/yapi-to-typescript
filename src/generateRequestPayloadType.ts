import { JSONSchema4 } from 'json-schema'
// @ts-ignore
import jsonSchemaGenerator from 'json-schema-generator'
import JSON5 from 'json5'
import { Api, Method, Required, RequestBodyType, RequestFormItemType, PropDefinition } from './types'
import propDefinitionsToJsonSchema from './propDefinitionsToJsonSchema'
import jsonSchemaToTypes from './jsonSchemaToTypes'

/**
 * 生成请求数据的 typescript 类型定义。
 *
 * @param api Api
 * @param interfaceName 接口名称
 * @returns typescript 定义
 */
export default async function generateRequestPayloadType(api: Api, interfaceName: string): Promise<string> {
  let jsonSchema: JSONSchema4 = {}
  switch (api.method) {
    case Method.GET:
    case Method.HEAD:
    case Method.OPTIONS:
      jsonSchema = propDefinitionsToJsonSchema(
        api.req_query.map<PropDefinition>(item => ({
          name: item.name,
          required: item.required === Required.True,
          type: 'string',
          comment: item.desc,
        })),
      )
      break
    default:
      switch (api.req_body_type) {
        case RequestBodyType.Form:
          jsonSchema = propDefinitionsToJsonSchema(
            api.req_body_form.map<PropDefinition>(item => ({
              name: item.name,
              required: item.required === Required.True,
              type: (item.type === RequestFormItemType.File ? 'file' : 'string') as any,
              comment: item.desc,
            })),
          )
          break
        case RequestBodyType.Json:
          if (api.req_body_other) {
            jsonSchema = api.req_body_is_json_schema
              ? JSON.parse(api.req_body_other)
              : jsonSchemaGenerator(JSON5.parse(api.req_body_other))
          }
          break
        default:
          break
      }
      break
  }
  return jsonSchemaToTypes(jsonSchema, interfaceName)
}
