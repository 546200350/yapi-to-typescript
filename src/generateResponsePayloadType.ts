import { JSONSchema4 } from 'json-schema'
import JSON5 from 'json5'
import { Api, ResponseBodyType } from './types'
import mockjsJsonToJsonSchema from './mockjsJsonToJsonSchema'
import jsonSchemaToTypes from './jsonSchemaToTypes'

/**
 * 生成返回数据的 typescript 类型定义。
 *
 * @param api Api
 * @param interfaceName 接口名称
 * @param [dataKey] 数据所在字段，不设置表示整体都是数据
 * @returns typescript 定义
 */
export default async function generateResponsePayloadType(api: Api, interfaceName: string, dataKey?: string): Promise<string> {
  let jsonSchema: JSONSchema4
  switch (api.res_body_type) {
    case ResponseBodyType.Json:
    case ResponseBodyType.JsonSchema:
      if (api.res_body_is_json_schema) {
        jsonSchema = api.res_body && JSON.parse(api.res_body)
      } else {
        jsonSchema = api.res_body && mockjsJsonToJsonSchema(JSON5.parse(api.res_body))
      }
      break
    default:
      break
  }
  if (dataKey && jsonSchema && jsonSchema.properties && jsonSchema.properties[dataKey]) {
    jsonSchema = jsonSchema.properties[dataKey]
  }
  return jsonSchemaToTypes(jsonSchema, interfaceName)
}
