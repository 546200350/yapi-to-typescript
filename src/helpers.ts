export class FileData<T = any> {
  /**
   * 原始文件数据。
   */
  private originalFileData: T

  /**
   * 文件数据辅助类，统一网页、小程序等平台的文件上传。
   *
   * @param originalFileData 原始文件数据
   */
  public constructor(originalFileData: T) {
    this.originalFileData = originalFileData
  }

  /**
   * 获取原始文件数据。
   *
   * @returns 原始文件数据
   */
  public getOriginalFileData(): T {
    return this.originalFileData
  }
}

/**
 * 解析请求数据，从请求数据中分离出普通数据和文件数据。
 *
 * @param [requestData] 要解析的请求数据
 * @returns 包含普通数据(data)和文件数据(fileData)的对象，data、fileData 为空对象时，表示没有此类数据
 */
export function parseRequestData(requestData?: any): { data: any, fileData: any } {
  const result = {
    data: {} as any,
    fileData: {} as any,
  }
  if (requestData != null) {
    if (typeof requestData === 'object' && !Array.isArray(requestData)) {
      Object.keys(requestData).forEach(key => {
        if (requestData[key] && requestData[key] instanceof FileData) {
          result.fileData[key] = (requestData[key] as FileData).getOriginalFileData()
        } else {
          result.data[key] = requestData[key]
        }
      })
    } else {
      result.data = requestData
    }
  }
  return result
}
