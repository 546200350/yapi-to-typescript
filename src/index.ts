import path from 'path'
import fs from 'fs-extra'
import * as changeCase from 'change-case'
import consola from 'consola'
import { Config, ApiList, InterfaceType, ExtendedApi } from './types'
import fetchApiCollection from './fetchApiCollection'
import generateRequestPayloadType from './generateRequestPayloadType'
import generateResponsePayloadType from './generateResponsePayloadType'

export default async (config: Config): Promise<void> => {
  consola.info('获取接口 JSON 文件中...')
  // 获取接口的 JSON 文件并转换为对象
  const apiCollection = await fetchApiCollection(config)
  consola.info('生成 TypeScript 类型文件中...')
  // 生成分类 ID 到分类 API 列表的对象
  const categoryIdToApiList = apiCollection.reduce<{ [id: number]: ApiList }>((res, api) => {
    if (api.list.length) {
      res[api.list[0].catid] = api.list
    }
    return res
  }, {})
  const tsContent = (
    await Promise.all(
      Object.keys(config.categories).map(async (categoryId: any) => {
        const { getRequestFunctionName, getInterfaceName } = config.categories[categoryId]
        return Promise.all(
          categoryIdToApiList[categoryId].map(async api => {
            const extendedApi: ExtendedApi = {
              ...api,
              parsedPath: path.parse(api.path),
              changeCase: changeCase,
            }
            const requestDataInterfaceName = changeCase.pascalCase(getInterfaceName(extendedApi, InterfaceType.Request))
            const responseDataInterfaceName = changeCase.pascalCase(getInterfaceName(extendedApi, InterfaceType.Response))
            const requestPayloadType = (await generateRequestPayloadType(api, requestDataInterfaceName)).trim()
            const responsePayloadType = (await generateResponsePayloadType(api, responseDataInterfaceName, config.dataKey)).trim()
            return [
              `/**\n * **请求类型**：${api.title}\n */\n${requestPayloadType}`,
              `/**\n * **响应类型**：${api.title}\n */\n${responsePayloadType}`,
              `/**\n * ${api.title}\n */\nexport function ${getRequestFunctionName(extendedApi)}(requestData${/(\{\}|any)$/s.test(requestPayloadType) ? '?' : ''}: ${requestDataInterfaceName}): Promise<${responseDataInterfaceName}> {\n${
                [
                  `  const { data, fileData } = parseRequestData(requestData)`,
                  `  return request({`,
                  `    path: '${api.path}',`,
                  `    method: '${api.method}',`,
                  `    requestBodyType: '${api.req_body_type}',`,
                  `    responseBodyType: '${api.res_body_type}',`,
                  `    data: data,`,
                  `    fileData: fileData`,
                  `  } as any)`,
                ].join('\n')
              }\n}`,
            ].join('\n\n')
          })
        )
      })
    )
  )
    .reduce((res, arr) => {
      res.push(...arr)
      return res
    }, [])
    .join('\n\n')
  const targetFile = path.resolve(process.cwd(), config.targetFile)
  const requestFile = path.join(path.parse(targetFile).dir, 'request.ts')
  if (!fs.existsSync(requestFile)) {
    fs.outputFileSync(requestFile, `${`
      import { RequestPayload } from 'yapi-to-typescript/lib/types'

      export default function request({
        path,
        method,
        requestBodyType,
        responseBodyType,
        data,
        fileData
      }: RequestPayload): Promise<any> {
        return new Promise((resolve, reject) => {
          // 是否含有文件数据
          const hasFileData = fileData !== null

          // 在这里实现请求逻辑
        })
      }
    `.trim().replace(/ {6}/g, '')}\n`)
  }
  fs.outputFileSync(targetFile, [
    `/* tslint:disable */\n/* eslint-disable */`,
    `import request from './request'`,
    `// @ts-ignore\nimport { FileData, parseRequestData } from 'yapi-to-typescript/lib/utils'`,
    tsContent,
  ].join('\n\n'))
  consola.success(`操作完成.`)
}
