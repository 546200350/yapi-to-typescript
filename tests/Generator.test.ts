import fs from 'fs-extra'
import path from 'path'
import tempy from 'tempy'
import { forOwn } from 'vtils'
import { Generator } from '../src/Generator'

const generatorFactory = (id: number | number[], typesOnly: boolean) => {
  const apiDir = tempy.directory()
  return new Generator({
    serverUrl: 'http://foo.bar',
    typesOnly: typesOnly,
    prodEnvName: 'production',
    outputFilePath: path.join(apiDir, 'index.ts'),
    requestFunctionFilePath: path.join(apiDir, 'request.ts'),
    projects: [
      {
        token: 'hello',
        categories: [
          {
            id: id,
            preproccessInterface(ii) {
              ii.path += '_test'
              return ii
            },
            getRequestFunctionName(ii, cc) {
              return cc.camelCase(ii.parsedPath.name)
            },
            getRequestDataTypeName(ii, cc) {
              return `${cc.pascalCase(ii.parsedPath.name)}Request`
            },
            getResponseDataTypeName(ii, cc) {
              return `${cc.pascalCase(ii.parsedPath.name)}Response`
            },
          },
        ],
      },
    ],
  })
}

describe('Generator', () => {
  test('正确生成代码并写入文件 - 单分类', async () => {
    const generator = generatorFactory(58, false)
    const output = await generator.generate()
    forOwn(output, ({ content }) => {
      expect(content).toMatchSnapshot('输出内容')
    })

    await generator.write(output)
    forOwn(output, ({ requestFilePath }, outputFilePath) => {
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot('接口文件')
      expect(fs.readFileSync(requestFilePath).toString()).toMatchSnapshot('请求文件')
    })
  })

  test('正确生成代码并写入文件 - 多分类', async () => {
    const generator = generatorFactory([58, 113], false)
    const output = await generator.generate()
    forOwn(output, ({ content }) => {
      expect(content).toMatchSnapshot('输出内容')
    })

    await generator.write(output)
    forOwn(output, ({ requestFilePath }, outputFilePath) => {
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot('接口文件')
      expect(fs.readFileSync(requestFilePath).toString()).toMatchSnapshot('请求文件')
    })
  })

  test('正确生成代码并写入文件 - 全部分类', async () => {
    const generator = generatorFactory(0, false)
    const output = await generator.generate()
    forOwn(output, ({ content }) => {
      expect(content).toMatchSnapshot('输出内容')
    })

    await generator.write(output)
    forOwn(output, ({ requestFilePath }, outputFilePath) => {
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot('接口文件')
      expect(fs.readFileSync(requestFilePath).toString()).toMatchSnapshot('请求文件')
    })
  })

  test('只生成类型代码并写入文件', async () => {
    const generator = generatorFactory(58, true)
    const output = await generator.generate()
    forOwn(output, ({ content }) => {
      expect(content).toMatchSnapshot('输出内容')
    })

    await generator.write(output)
    forOwn(output, ({ requestFilePath }, outputFilePath) => {
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot('接口文件')
      expect(fs.existsSync(requestFilePath)).toBe(false)
    })
  })
})
