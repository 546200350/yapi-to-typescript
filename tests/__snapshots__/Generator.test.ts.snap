// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator 正确生成代码并写入文件 1`] = `
Array [
  "const mockUrl_0_0_0 = \\"http://foo.bar/mock/32\\"
const prodUrl_0_0_0 = \\"http://prod.com/\\"
const dataKey_0_0_0 = undefined


/**
 * 接口 **5555** 的 **请求类型**
 */
export interface PutTestRequest {
  /**
   * ID
   */
  id: string;
}

/**
 * 接口 **5555** 的 **返回类型**
 */
export interface PutTestResponse {}

/**
 * 接口 **5555** 的 **请求函数**
 */
export function putTest(requestData: PutTestRequest): Promise<PutTestResponse> {
  return request({
    ...putTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **5555** 的 **请求配置**
 */
putTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/put_test\\",
  method: Method.PUT,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.raw,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/put_test\\",
  undefined
>)


/**
 * 接口 **get** 的 **请求类型**
 */
export interface GetTestRequest {}

/**
 * 接口 **get** 的 **返回类型**
 */
export interface GetTestResponse {
  data: string;
}

/**
 * 接口 **get** 的 **请求函数**
 */
export function getTest(requestData?: GetTestRequest): Promise<GetTestResponse> {
  return request({
    ...getTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **get** 的 **请求配置**
 */
getTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/get_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/get_test\\",
  undefined
>)


/**
 * 接口 **json5** 的 **请求类型**
 */
export interface Json5TestRequest {}

/**
 * 接口 **json5** 的 **返回类型**
 */
export interface Json5TestResponse {
  id?: number;
  age?: string;
  name?: string;
}

/**
 * 接口 **json5** 的 **请求函数**
 */
export function json5Test(requestData?: Json5TestRequest): Promise<Json5TestResponse> {
  return request({
    ...json5Test.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **json5** 的 **请求配置**
 */
json5Test.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/json5_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/json5_test\\",
  undefined
>)


/**
 * 接口 **json5-Request** 的 **请求类型**
 */
export interface JsonRequestTestRequest {
  id: number;
  name: string;
  likes: {}[];
}

/**
 * 接口 **json5-Request** 的 **返回类型**
 */
export interface JsonRequestTestResponse {}

/**
 * 接口 **json5-Request** 的 **请求函数**
 */
export function jsonRequestTest(requestData: JsonRequestTestRequest): Promise<JsonRequestTestResponse> {
  return request({
    ...jsonRequestTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **json5-Request** 的 **请求配置**
 */
jsonRequestTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/jsonRequest_test\\",
  method: Method.POST,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/jsonRequest_test\\",
  undefined
>)


/**
 * 接口 **post** 的 **请求类型**
 */
export interface PostTestRequest {
  page: number;
}

/**
 * 接口 **post** 的 **返回类型**
 */
export interface PostTestResponse {
  list: string[];
}

/**
 * 接口 **post** 的 **请求函数**
 */
export function postTest(requestData: PostTestRequest): Promise<PostTestResponse> {
  return request({
    ...postTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **post** 的 **请求配置**
 */
postTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/post_test\\",
  method: Method.POST,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/post_test\\",
  undefined
>)


/**
 * 接口 **x** 的 **请求类型**
 */
export interface XxTestRequest {
  id: string;
  namre?: string;
}

/**
 * 接口 **x** 的 **返回类型**
 */
export interface XxTestResponse {
  /**
   * 错误
   */
  err: number;
  /**
   * 错误详情
   */
  msg: string;
  /**
   * 数据
   */
  data?: {};
}

/**
 * 接口 **x** 的 **请求函数**
 */
export function xxTest(requestData: XxTestRequest): Promise<XxTestResponse> {
  return request({
    ...xxTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **x** 的 **请求配置**
 */
xxTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/xx_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/xx_test\\",
  undefined
>)


/**
 * 接口 **没返回数据** 的 **请求类型**
 */
export interface NoResponseDataTestRequest {}

/**
 * 接口 **没返回数据** 的 **返回类型**
 */
export interface NoResponseDataTestResponse {}

/**
 * 接口 **没返回数据** 的 **请求函数**
 */
export function noResponseDataTest(requestData?: NoResponseDataTestRequest): Promise<NoResponseDataTestResponse> {
  return request({
    ...noResponseDataTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **没返回数据** 的 **请求配置**
 */
noResponseDataTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/noResponseData_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/noResponseData_test\\",
  undefined
>)",
]
`;

exports[`Generator 正确生成代码并写入文件 2`] = `
"/* tslint:disable */
/* eslint-disable */

/**
 * **该文件由 yapi-to-typescript 自动生成，请勿直接修改！！！** 
 */

import request from \\"./request\\"

// @ts-ignore
import { Method, RequestBodyType, ResponseBodyType, RequestConfig, FileData, parseRequestData } from 'yapi-to-typescript'

const mockUrl_0_0_0 = \\"http://foo.bar/mock/32\\"
const prodUrl_0_0_0 = \\"http://prod.com/\\"
const dataKey_0_0_0 = undefined


/**
 * 接口 **5555** 的 **请求类型**
 */
export interface PutTestRequest {
  /**
   * ID
   */
  id: string;
}

/**
 * 接口 **5555** 的 **返回类型**
 */
export interface PutTestResponse {}

/**
 * 接口 **5555** 的 **请求函数**
 */
export function putTest(requestData: PutTestRequest): Promise<PutTestResponse> {
  return request({
    ...putTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **5555** 的 **请求配置**
 */
putTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/put_test\\",
  method: Method.PUT,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.raw,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/put_test\\",
  undefined
>)


/**
 * 接口 **get** 的 **请求类型**
 */
export interface GetTestRequest {}

/**
 * 接口 **get** 的 **返回类型**
 */
export interface GetTestResponse {
  data: string;
}

/**
 * 接口 **get** 的 **请求函数**
 */
export function getTest(requestData?: GetTestRequest): Promise<GetTestResponse> {
  return request({
    ...getTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **get** 的 **请求配置**
 */
getTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/get_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/get_test\\",
  undefined
>)


/**
 * 接口 **json5** 的 **请求类型**
 */
export interface Json5TestRequest {}

/**
 * 接口 **json5** 的 **返回类型**
 */
export interface Json5TestResponse {
  id?: number;
  age?: string;
  name?: string;
}

/**
 * 接口 **json5** 的 **请求函数**
 */
export function json5Test(requestData?: Json5TestRequest): Promise<Json5TestResponse> {
  return request({
    ...json5Test.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **json5** 的 **请求配置**
 */
json5Test.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/json5_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/json5_test\\",
  undefined
>)


/**
 * 接口 **json5-Request** 的 **请求类型**
 */
export interface JsonRequestTestRequest {
  id: number;
  name: string;
  likes: {}[];
}

/**
 * 接口 **json5-Request** 的 **返回类型**
 */
export interface JsonRequestTestResponse {}

/**
 * 接口 **json5-Request** 的 **请求函数**
 */
export function jsonRequestTest(requestData: JsonRequestTestRequest): Promise<JsonRequestTestResponse> {
  return request({
    ...jsonRequestTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **json5-Request** 的 **请求配置**
 */
jsonRequestTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/jsonRequest_test\\",
  method: Method.POST,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/jsonRequest_test\\",
  undefined
>)


/**
 * 接口 **post** 的 **请求类型**
 */
export interface PostTestRequest {
  page: number;
}

/**
 * 接口 **post** 的 **返回类型**
 */
export interface PostTestResponse {
  list: string[];
}

/**
 * 接口 **post** 的 **请求函数**
 */
export function postTest(requestData: PostTestRequest): Promise<PostTestResponse> {
  return request({
    ...postTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **post** 的 **请求配置**
 */
postTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/post_test\\",
  method: Method.POST,
  requestBodyType: RequestBodyType.json,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/post_test\\",
  undefined
>)


/**
 * 接口 **x** 的 **请求类型**
 */
export interface XxTestRequest {
  id: string;
  namre?: string;
}

/**
 * 接口 **x** 的 **返回类型**
 */
export interface XxTestResponse {
  /**
   * 错误
   */
  err: number;
  /**
   * 错误详情
   */
  msg: string;
  /**
   * 数据
   */
  data?: {};
}

/**
 * 接口 **x** 的 **请求函数**
 */
export function xxTest(requestData: XxTestRequest): Promise<XxTestResponse> {
  return request({
    ...xxTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **x** 的 **请求配置**
 */
xxTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/xx_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/xx_test\\",
  undefined
>)


/**
 * 接口 **没返回数据** 的 **请求类型**
 */
export interface NoResponseDataTestRequest {}

/**
 * 接口 **没返回数据** 的 **返回类型**
 */
export interface NoResponseDataTestResponse {}

/**
 * 接口 **没返回数据** 的 **请求函数**
 */
export function noResponseDataTest(requestData?: NoResponseDataTestRequest): Promise<NoResponseDataTestResponse> {
  return request({
    ...noResponseDataTest.requestConfig,
    ...parseRequestData(requestData)
  } as any)
}

/**
 * 接口 **没返回数据** 的 **请求配置**
 */
noResponseDataTest.requestConfig = Object.freeze({
  mockUrl: mockUrl_0_0_0,
  prodUrl: prodUrl_0_0_0,
  path: \\"/noResponseData_test\\",
  method: Method.GET,
  requestBodyType: RequestBodyType.query,
  responseBodyType: ResponseBodyType.json,
  dataKey: dataKey_0_0_0
} as RequestConfig<
  \\"http://foo.bar/mock/32\\",
  \\"http://prod.com/\\",
  \\"/noResponseData_test\\",
  undefined
>)
"
`;

exports[`Generator 正确生成代码并写入文件 3`] = `
"import { RequestFunction } from 'yapi-to-typescript'

/** 是否是生产环境 */
const isProd = false

/**
 * 请求函数。
 *
 * **注意**：若 dataKey 不为空，取得接口返回值后，应类似这样返回结果：
 *
 * \`\`\`js
 * return dataKey ? (response[dataKey] || response) : response
 * \`\`\`
 */
const request: RequestFunction = ({
  /** 接口 Mock 地址，结尾无 \`/\` */
  mockUrl,
  /** 接口生产环境地址，结尾无 \`/\` */
  prodUrl,
  /** 接口路径，以 \`/\` 开头 */
  path,
  /** 请求方法 */
  method,
  /** 请求数据类型 */
  requestBodyType,
  /** 返回数据类型 */
  responseBodyType,
  /** 接口返回值中数据所在的键 */
  dataKey,
  /** 请求数据，不含文件数据 */
  data,
  /** 请求文件数据 */
  fileData
}): Promise<any> => {
  return new Promise((resolve, reject) => {
    /** 请求地址 */
    const url = \`\${isProd ? prodUrl : mockUrl}\${path}\`

    /** 是否含有文件数据 */
    const hasFileData = Object.keys(fileData).length > 0

    // 在这里实现请求逻辑
  })
}

export default request
"
`;
